packages/strapi-database/lib/database-manager.js

```diff
@@ -16,6 +16,59 @@ class DatabaseManager {
     this.models = new Map();
   }
 
+  // Check if all collection names are unique
+  checkForDuplicates() {
+    const createErrorMessage = (
+      collectionA,
+      collectionB
+    ) => `Duplicated collection name: \`${collectionA.collectionName}\`.
+The same collection name can't be used for two different models.
+First found in ${collectionA.origin} \`${collectionA.apiOrPluginName}\`, model \`${collectionA.modelName}\`.
+Second found in ${collectionB.origin} \`${collectionB.apiOrPluginName}\`, model \`${collectionB.modelName}\`.`;
+
+    const collections = [];
+    _.forIn(this.strapi.admin.models, (model, modelName) => {
+      collections.push({
+        origin: 'Strapi internal',
+        collectionName: model.collectionName || `${modelName}`.toLocaleLowerCase(),
+        apiOrPluginName: 'admin',
+        modelName,
+      });
+    });
+
+    _.forIn(this.strapi.api, (api, apiName) => {
+      _.forIn(api.models, (model, modelName) => {
+        collections.push({
+          origin: 'API',
+          collectionName: model.collectionName || `${modelName}`.toLocaleLowerCase(),
+          apiOrPluginName: apiName,
+          modelName,
+        });
+      });
+    });
+
+    _.forIn(this.strapi.plugins, (plugin, pluginName) => {
+      _.forIn(plugin.models, (model, modelName) => {
+        collections.push({
+          origin: 'Plugin',
+          collectionName: model.collectionName || `${modelName}`.toLocaleLowerCase(),
+          apiOrPluginName: pluginName,
+          modelName,
+        });
+      });
+    });
+
+    collections.forEach((collectionA, indexA) => {
+      const similarCollectionFound = collections
+        .slice(indexA + 1)
+        .find(colB => colB.collectionName === collectionA.collectionName);
```

Comment:
you can avoid slicing the array by comparing colA.collectionName == colB.collectionName && colA.uid != colB.uid. This will avoid duplicating the array on each iteration